File: react-native/src/gk1.ts
// GK1 wire codec: "GK1." + base64url(XOR(utf8(json), key)). Obfuscation + a stable
// envelope, not encryption (HTTPS does transit security; host-pinned URLs stop redirects).
// Byte-identical to the Dart/Python implementations - do not change the tag or key.
const KEY = 'pink-rorqual//golden-krill//v1';
const TAG = 'GK1.';

const B64 = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_';

function b64urlEncode(bytes: Uint8Array): string {
  let out = '';
  for (let i = 0; i < bytes.length; i += 3) {
    const b0 = bytes[i];
    const b1 = i + 1 < bytes.length ? bytes[i + 1] : 0;
    const b2 = i + 2 < bytes.length ? bytes[i + 2] : 0;
    out += B64[b0 >> 2];
    out += B64[((b0 & 3) << 4) | (b1 >> 4)];
    if (i + 1 < bytes.length) out += B64[((b1 & 15) << 2) | (b2 >> 6)];
    if (i + 2 < bytes.length) out += B64[b2 & 63];
  }
  while (out.length % 4 !== 0) out += '='; // url-safe WITH padding, matching Dart/Python
  return out;
}

function b64urlDecode(s: string): Uint8Array {
  const lookup: Record<string, number> = {};
  for (let i = 0; i < B64.length; i++) lookup[B64[i]] = i;
  const clean = s.replace(/=+$/, '');
  const out: number[] = [];
  for (let i = 0; i < clean.length; i += 4) {
    const c0 = lookup[clean[i]];
    const c1 = lookup[clean[i + 1]];
    const c2 = lookup[clean[i + 2]];
    const c3 = lookup[clean[i + 3]];
    out.push((c0 << 2) | (c1 >> 4));
    if (clean[i + 2] !== undefined) out.push(((c1 & 15) << 4) | (c2 >> 2));
    if (clean[i + 3] !== undefined) out.push(((c2 & 3) << 6) | c3);
  }
  return new Uint8Array(out);
}

function utf8Encode(s: string): Uint8Array {
  // TextEncoder is available in Hermes/modern RN; fall back to a manual encoder.
  if (typeof TextEncoder !== 'undefined') return new TextEncoder().encode(s);
  const out: number[] = [];
  for (let i = 0; i < s.length; i++) {
    let c = s.charCodeAt(i);
    if (c < 0x80) out.push(c);
    else if (c < 0x800) out.push(0xc0 | (c >> 6), 0x80 | (c & 0x3f));
    else out.push(0xe0 | (c >> 12), 0x80 | ((c >> 6) & 0x3f), 0x80 | (c & 0x3f));
  }
  return new Uint8Array(out);
}

function utf8Decode(bytes: Uint8Array): string {
  if (typeof TextDecoder !== 'undefined') return new TextDecoder().decode(bytes);
  let out = '';
  for (let i = 0; i < bytes.length; ) {
    const b = bytes[i++];
    if (b < 0x80) out += String.fromCharCode(b);
    else if (b < 0xe0) out += String.fromCharCode(((b & 0x1f) << 6) | (bytes[i++] & 0x3f));
    else out += String.fromCharCode(((b & 0x0f) << 12) | ((bytes[i++] & 0x3f) << 6) | (bytes[i++] & 0x3f));
  }
  return out;
}

function xor(bytes: Uint8Array): Uint8Array {
  const k = utf8Encode(KEY);
  const out = new Uint8Array(bytes.length);
  for (let i = 0; i < bytes.length; i++) out[i] = bytes[i] ^ k[i % k.length];
  return out;
}

/** Encode a JSON string into a GK1 blob. */
export function encode(json: string): string {
  return TAG + b64urlEncode(xor(utf8Encode(json)));
}

/** Decode a GK1 blob to its JSON string. Tolerates a raw (non-GK1) JSON body. */
export function decode(blob: string): string {
  if (!blob.startsWith(TAG)) return blob;
  return utf8Decode(xor(b64urlDecode(blob.slice(TAG.length))));
}

File: unity/Runtime/Gk1.cs
using System.Collections.Generic;
using System.Text;

namespace GoldenKrill
{
    /// GK1 wire codec: "GK1." + base64url(XOR(utf8(json), key)). Obfuscation + a stable
    /// envelope, not encryption (HTTPS does transit security; host-pinned URLs stop redirects).
    /// Byte-identical to the Dart / Python / TS implementations: do not change the tag or key.
    /// The conformance vector (Tests/gk1_vectors.json) pins this against every other port.
    public static class Gk1
    {
        private const string Key = "pink-rorqual//golden-krill//v1";
        private const string Tag = "GK1.";

        // URL-safe alphabet (-, _) and WITH '=' padding, matching Dart/Python/TS exactly.
        private const string B64 = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_";

        private static string B64UrlEncode(byte[] bytes)
        {
            var sb = new StringBuilder();
            for (int i = 0; i < bytes.Length; i += 3)
            {
                int b0 = bytes[i];
                int b1 = i + 1 < bytes.Length ? bytes[i + 1] : 0;
                int b2 = i + 2 < bytes.Length ? bytes[i + 2] : 0;
                sb.Append(B64[b0 >> 2]);
                sb.Append(B64[((b0 & 3) << 4) | (b1 >> 4)]);
                if (i + 1 < bytes.Length) sb.Append(B64[((b1 & 15) << 2) | (b2 >> 6)]);
                if (i + 2 < bytes.Length) sb.Append(B64[b2 & 63]);
            }
            while (sb.Length % 4 != 0) sb.Append('='); // url-safe WITH padding
            return sb.ToString();
        }

        private static byte[] B64UrlDecode(string s)
        {
            var lookup = new Dictionary<char, int>();
            for (int i = 0; i < B64.Length; i++) lookup[B64[i]] = i;
            string clean = s.TrimEnd('=');
            var outBytes = new List<byte>();
            for (int i = 0; i < clean.Length; i += 4)
            {
                int c0 = lookup[clean[i]];
                int c1 = lookup[clean[i + 1]];
                bool has2 = i + 2 < clean.Length;
                bool has3 = i + 3 < clean.Length;
                int c2 = has2 ? lookup[clean[i + 2]] : 0;
                int c3 = has3 ? lookup[clean[i + 3]] : 0;
                outBytes.Add((byte)((c0 << 2) | (c1 >> 4)));
                if (has2) outBytes.Add((byte)(((c1 & 15) << 4) | (c2 >> 2)));
                if (has3) outBytes.Add((byte)(((c2 & 3) << 6) | c3));
            }
            return outBytes.ToArray();
        }

        private static byte[] Xor(byte[] bytes)
        {
            byte[] k = Encoding.UTF8.GetBytes(Key);
            var outBytes = new byte[bytes.Length];
            for (int i = 0; i < bytes.Length; i++) outBytes[i] = (byte)(bytes[i] ^ k[i % k.Length]);
            return outBytes;
        }

        /// Encode a JSON string into a GK1 blob.
        public static string Encode(string json)
        {
            return Tag + B64UrlEncode(Xor(Encoding.UTF8.GetBytes(json)));
        }

        /// Decode a GK1 blob to its JSON string. Tolerates a raw (non-GK1) JSON body.
        public static string Decode(string blob)
        {
            if (blob == null || !blob.StartsWith(Tag)) return blob;
            return Encoding.UTF8.GetString(Xor(B64UrlDecode(blob.Substring(Tag.Length))));
        }
    }
}
